leaders/src/main/java/edu/cornell/worker/DockerWorker.java

```diff
@@ -0,0 +1,159 @@
+package edu.cornell.worker;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.async.ResultCallback;
+import com.github.dockerjava.api.model.HostConfig;
+import com.github.dockerjava.core.DefaultDockerClientConfig;
+import com.github.dockerjava.core.DockerClientConfig;
+import com.github.dockerjava.core.DockerClientImpl;
+import com.github.dockerjava.httpclient5.ApacheDockerHttpClient;
+import com.github.dockerjava.transport.DockerHttpClient;
+import lombok.EqualsAndHashCode;
+import lombok.NonNull;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Set;
+
+import static edu.cornell.Main.*;
+
+/** A worker that runs inside a Docker container */
+@Slf4j
+@EqualsAndHashCode
+@ToString
+public class DockerWorker implements Worker {
+
+    /**
+     * The link to the workers Docker image
+     */
+    private static final String IMAGE_URL = "ghcr.io/jakejack13/speed-workers:latest";
+
+    /**
+     * The Docker http client
+     */
+    private final DockerHttpClient httpClient;
+
+    /**
+     * The Docker client
+     */
+    private final DockerClient dockerClient;
+    /**
+     * The id of the container running the worker
+     */
+    private final String id;
+
+    /**
+     * true if the worker has finished running, false otherwise
+     */
+    private boolean done = false;
+
+    /**
+     * Creates a new DockerWorker
+     * @param repoUrl the url of the repository
+     * @param repoBranch the branch of the repository to test
+     * @param tests the set of tests to run
+     * @param kafkaAddress the address of the Kafka message bus to send test results to
+     */
+    DockerWorker(@NonNull String repoUrl, @NonNull String repoBranch, @NonNull Set<String> tests,
+            @NonNull String kafkaAddress) {
+        DockerClientConfig config =
+                DefaultDockerClientConfig.createDefaultConfigBuilder()
+                        .build();
+
+        httpClient = new ApacheDockerHttpClient.Builder()
+                .dockerHost(config.getDockerHost())
+                .sslConfig(config.getSSLConfig())
+                .maxConnections(100)
+                .connectionTimeout(Duration.ofSeconds(30))
+                .responseTimeout(Duration.ofSeconds(45))
+                .build();
+
+        dockerClient = DockerClientImpl.getInstance(config, httpClient);
+
+        HostConfig hostConfig = new HostConfig()
+                .withNetworkMode("kafka_default"); //default kafka network name
+
+        id = dockerClient.createContainerCmd(IMAGE_URL)
+                .withPortSpecs("29092:29092")
+                .withEnv(ENV_REPO_URL + '=' + repoUrl,
+                        ENV_REPO_BRANCH + '=' + repoBranch,
+                        ENV_REPO_TESTS + '=' + String.join(",", tests),
+                        ENV_KAFKA_ADDRESS + '=' + kafkaAddress)
+                .withHostConfig(hostConfig)
```

Comment:
currently, the kafkaAddress here is useless until structure is changed to use "kafka:29092" instead of the same thing the leader uses. If/When the leader is moved to using a docker container, the leader will need to use "kafka:29092" as well and be placed into the kafka_default network.

Reply:
Should we move the `kafka` docker-compose file to top-level and add the leader to it? That way, everything boots up at the same time. We can configure the network on there

Reply:
Let's experiment with this in a different PR because what we have works. Since all of these services depend on one another booting up our current system we know it works because we can wait for Kafka to be fully up then the leader etc. Although I suspect it should be fine to make this change, I think we should keep this working copy and then make that change separately.